File: pcm/matlab/pcm.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% A律非均匀PCM编译码程序
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
clc;
clear;

fs=32000;
LEN_SIG=1024;
t=1/fs*(0:LEN_SIG-1);

SCALE_SHIFT=10;
SCALE_PCM=2^SCALE_SHIFT-1;

LEN_VIEW=256;
View_X=[1:LEN_VIEW];

%---------------------------------------------------
%   Source Data
%---------------------------------------------------
SrcSignal=round(SCALE_PCM*sin(2*pi*800*t));  %  800Hz;
figure(1);
subplot(2,1,1);plot(View_X,SrcSignal(View_X));
axis([0 LEN_VIEW-1 -SCALE_PCM SCALE_PCM]);
title('输入信号波形');

%---------------------------------------------------
%   A律非均匀PCM编码
%---------------------------------------------------
for i = 1 : LEN_SIG
    if SrcSignal(i) >= 0
        out(i,1) = 0 ;
    else 
        out(i,1) = 1 ;
    end
        
    if abs(SrcSignal(i)) >= 0 & abs(SrcSignal(i)) < 16
        out(i,2) = 0;
        out(i,3) = 0;
        out(i,4) = 0;
        step = 1;
        st = 0;
    elseif 16 <= abs(SrcSignal(i)) & abs(SrcSignal(i)) < 32 
        out(i,2) = 0;
        out(i,3) = 0;
        out(i,4) = 1;
        step = 1;
        st = 16;
    elseif 32 <= abs(SrcSignal(i)) & abs(SrcSignal(i)) < 64 
        out(i,2) = 0;
        out(i,3) = 1;
        out(i,4) = 0;
        step = 2;
        st = 32;
    elseif 64 <= abs(SrcSignal(i)) & abs(SrcSignal(i)) < 128 
        out(i,2) = 0;
        out(i,3) = 1;
        out(i,4) = 1;
        step = 4;
        st = 64;
    elseif 128 <= abs(SrcSignal(i)) & abs(SrcSignal(i)) < 256 
        out(i,2) = 1;
        out(i,3) = 0;
        out(i,4) = 0;
        step = 8;
        st = 128;
    elseif 256 <= abs(SrcSignal(i)) & abs(SrcSignal(i)) < 512
        out(i,2) = 1;
        out(i,3) = 0;
        out(i,4) = 1;
        step = 16;
        st = 256;
    elseif 512 <= abs(SrcSignal(i)) & abs(SrcSignal(i)) < 1024
        out(i,2) = 1;
        out(i,3) = 1;
        out(i,4) = 0;
        step = 32;
        st = 512;
    elseif 1024 <= abs(SrcSignal(i)) & abs(SrcSignal(i)) < 2048
        out(i,2) = 1;
        out(i,3) = 1;
        out(i,4) = 1;
        step = 64;
        st = 1024;
    else
        out(i,2) = 1;
        out(i,3) = 1;
        out(i,4) = 1;
        step = 64;
        st = 1024;
    end

    if abs(SrcSignal(i)) > 2048
        out(i,2:8) = [1 1 1 1 1 1 1];
    else
        tmp = floor((abs(SrcSignal(i)) - st)/step);
        t = dec2bin(tmp,4) - 48;
        out(i,5:8) = t(1:4)
    end
end
EncOut = reshape(out',1,8*LEN_SIG);

% 打印编码数据
% fid_enc = fopen('PCM_enc.txt','W');
% 
% for j = 1 : LEN_SIG
%     x = out(j,1)*128 + out(j,2)*64 + out(j,3)*32 + out(j,4)*16 + out(j,5)*8 + out(j,6)*4 + out(j,7)*2 + out(j,8);
%     fprintf(fid_enc,"%d\n",x);
% end
% 
% fclose(fid_enc);

%---------------------------------------------------
%   A律非均匀PCM译码
%---------------------------------------------------
in = reshape(EncOut',8,LEN_SIG)';
slot(1) = 0;
slot(2) = 16;
slot(3) = 32;
slot(4) = 64;
slot(5) = 128;
slot(6) = 256;
slot(7) = 512;
slot(8) = 1024;

step(1) = 1;
step(2) = 1;
step(3) = 2;
step(4) = 4;
step(5) = 8;
step(6) = 16;
step(7) = 32;
step(8) = 64;

for i = 1 : LEN_SIG
    if in(i,1) == 0
        ss = 1;
    else
        ss = -1;
    end

    tmp = in(i,2) * 4 + in(i,3) * 2 + in(i,4) + 1;
    st = slot(tmp);
    dt = (in(i,5) * 8 + in(i,6) * 4 + in(i,7) * 2 + in(i,8)) * step(tmp);
    DecOut(i) = ss * (st + dt);
end

% 打印解码数据
% fid_dec = fopen('PCM_dec.txt','W');
% for j = 1 : LEN_SIG
%     fprintf(fid_dec,"%d\n",DecOut(j));
% end
% fclose(fid_dec);

subplot(2,1,2);plot(View_X,DecOut(View_X));
axis([0 LEN_VIEW-1 -SCALE_PCM SCALE_PCM]);
title('译码信号波形')

%---------------------------------------------------
%           CCS_DSP Export 
%---------------------------------------------------

fid1 = fopen('PCM_SigIn.txt','W');
fprintf(fid1,'PCM_SigIn[%d] = { \r',LEN_VIEW);

for j = 1 : LEN_VIEW
    x = floor(SrcSignal(j));
    if(x >= 32767.0)
        x = 32767 ;
    elseif(x < -32768.0)
        x = - 32768 ;
    end

    fprintf(fid1,"%6d,",x);
    if((mod(j,8) == 0) && (j > 1))
        fprintf(fid1,'\r');
    end
end

fclose(fid1);

File: pcm/c/pcm.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

short seg[8] = {16,32,64,128,256,512,1024,2048} ;
short slot[8] = {0,16,32,64,128,256,512,1024};
short level[8] = {1,1,2,4,8,16,32,64} ;
short datain,step,st,ss,dt;
short EncOut[256][8];
short DecOut[256];
short datarec[8];
short i,j;
short temp;
short PCM_SigIn[256]={
            0,   160,   316,   464,   601,   723,   828,   911, 
            973,  1010,  1023,  1010,   973,   911,   828,   723,
            601,   464,   316,   160,     0,  -160,  -316,  -464,
            -601,  -723,  -828,  -911,  -973, -1010, -1023, -1010,
            -973,  -911,  -828,  -723,  -601,  -464,  -316,  -160,
               0,   160,   316,   464,   601,   723,   828,   911,
             973,  1010,  1023,  1010,   973,   911,   828,   723,
             601,   464,   316,   160,     0,  -160,  -316,  -464,
            -601,  -723,  -828,  -911,  -973, -1010, -1023, -1010,
            -973,  -911,  -828,  -723,  -601,  -464,  -316,  -160,
               0,   160,   316,   464,   601,   723,   828,   911,
             973,  1010,  1023,  1010,   973,   911,   828,   723,
             601,   464,   316,   160,     0,  -160,  -316,  -464,
            -601,  -723,  -828,  -911,  -973, -1010, -1023, -1010,
            -973,  -911,  -828,  -723,  -601,  -464,  -316,  -160,
               0,   160,   316,   464,   601,   723,   828,   911,
             973,  1010,  1023,  1010,   973,   911,   828,   723,
             601,   464,   316,   160,     0,  -160,  -316,  -464,
            -601,  -723,  -828,  -911,  -973, -1010, -1023, -1010,
            -973,  -911,  -828,  -723,  -601,  -464,  -316,  -160,
               0,   160,   316,   464,   601,   723,   828,   911,
             973,  1010,  1023,  1010,   973,   911,   828,   723,
             601,   464,   316,   160,     0,  -160,  -316,  -464,
            -601,  -723,  -828,  -911,  -973, -1010, -1023, -1010,
            -973,  -911,  -828,  -723,  -601,  -464,  -316,  -160,
               0,   160,   316,   464,   601,   723,   828,   911,
             973,  1010,  1023,  1010,   973,   911,   828,   723,
             601,   464,   316,   160,     0,  -160,  -316,  -464,
            -601,  -723,  -828,  -911,  -973, -1010, -1023, -1010,
            -973,  -911,  -828,  -723,  -601,  -464,  -316,  -160,
               0,   160,   316,   464,   601,   723,   828,   911,
             973,  1010,  1023,  1010,   973,   911,   828,   723 
};


/*-----------------------------------
        Functions Declaration
-------------------------------------*/

void PCM_Encode(void);
void PCM_Decode(void);

void main()
{
    PCM_Encode();
    PCM_Decode();
    exit(0);
    do 
    {
    }while(1);
}

void PCM_Encode(void)
{
    for(i = 0 ; i < 256 ; i ++)
    {
        datain = PCM_SigIn[i];
        if(datain >= 0)
        {
            EncOut[i][0] = 0 ;
        }
        else
        {
            EncOut[i][0] = 1 ;
        }

        if((abs(datain) >= 0) && (abs(datain) < 16))
        {
            EncOut[i][1] = 0 ;
            EncOut[i][2] = 0 ;
            EncOut[i][3] = 0 ;
            step = 1 ;
            st = 0 ;
        }
        else if((abs(datain) >= 16) && (abs(datain) < 32))
        {
            EncOut[i][1] = 0 ;
            EncOut[i][2] = 0 ;
            EncOut[i][3] = 1 ;
            step = 1 ;
            st = 16 ;
        }
        else if((abs(datain) >= 32) && (abs(datain) < 64))
        {
            EncOut[i][1] = 0 ;
            EncOut[i][2] = 1 ;
            EncOut[i][3] = 0 ;
            step = 2 ;
            st = 32 ;
        }
        else if((abs(datain) >= 64) && (abs(datain) < 128))
        {
            EncOut[i][1] = 0 ;
            EncOut[i][2] = 1 ;
            EncOut[i][3] = 1 ;
            step = 4 ;
            st = 64 ;
        }
        else if((abs(datain) >= 128) && (abs(datain) < 256))
        {
            EncOut[i][1] = 1 ;
            EncOut[i][2] = 0 ;
            EncOut[i][3] = 0 ;
            step = 8 ;
            st = 128 ;
        }
        else if((abs(datain) >= 256) && (abs(datain) < 512))
        {
            EncOut[i][1] = 1 ;
            EncOut[i][2] = 0 ;
            EncOut[i][3] = 1 ;
            step = 16 ;
            st = 256 ;
        }
        else if((abs(datain) >= 512) && (abs(datain) < 1024))
        {
            EncOut[i][1] = 1 ;
            EncOut[i][2] = 1 ;
            EncOut[i][3] = 0 ;
            step = 32 ;
            st = 512 ;
        }
        else if((abs(datain) >= 1024) && (abs(datain) < 2048))
        {
            EncOut[i][1] = 1 ;
            EncOut[i][2] = 1 ;
            EncOut[i][3] = 1 ;
            step = 64 ;
            st = 1024 ;
        }
        else 
        {
            EncOut[i][1] = 1 ;
            EncOut[i][2] = 1 ;
            EncOut[i][3] = 1 ;
            step = 64 ;
            st = 1024 ;
        }

        if(abs(datain) > 2048)
        {
            for(j = 1 ; j < 7 ; j ++)
            {
                EncOut[i][j] = 1;
            }
        }

        temp = (abs(datain) - st)/step;

        EncOut[i][4] = temp >> 3;
        EncOut[i][5] = (temp - EncOut[i][4]<<3) >> 2;
        EncOut[i][6] = (temp - (EncOut[i][4]<<3) - (EncOut[i][5]<<2)) >> 1;
        EncOut[i][7] = (temp - (EncOut[i][4]<<3) - (EncOut[i][5]<<2) - (EncOut[i][6]<<1));

        //printf("%02d\n",EncOut[i][0] * 128 + EncOut[i][1] * 64 + EncOut[i][2] * 32 + EncOut[i][3] * 16 + EncOut[i][4] * 8 + EncOut[i][5] * 4 + EncOut[i][6] * 2 + EncOut[i][7] * 1);
    }
}

void PCM_Decode(void)
{
    for( i = 0 ; i < 256 ; i ++)
    {
        for( j = 0 ; j < 8 ; j ++)
        {
            datarec[j] = EncOut[i][j];
        }

        if(datarec[0] == 0)
            ss = 1;
        else
            ss = - 1;

        temp = datarec[1] * 4 + datarec[2] * 2 + datarec[3];
        st = slot[temp];
        dt = (datarec[4] * 8 + datarec[5] * 4 + datarec[6] * 2 + datarec[7]) * level[temp];
        DecOut[i] = ss * (st + dt);

        //printf("%d\n",DecOut[i]);
    }
}
